### ISSUE TITLE
False positive ``W0612`` / ``unused-variable`` with dataclasses

### ISSUE THREAD

**Issue by DervishD** [pylint-dev/pylint#9258]
### Bug description

With the minimal code shown below, I get error W0612, unused variable _HAS_DEFAULT_FACTORY at line 1 column 1, the start of the code. The module docstring is not relevant, it is just there to show that the error is signalled at line 1, column 1, no matter the content.

My config file forces all warnings, that's true, but it is still a false positive.

```python
"""."""

import dataclasses

@dataclasses.dataclass
class SomeDataClass():
    """."""

SomeDataClass()
```

### Configuration

```ini
[MASTER]
jobs=0
unsafe-load-any-extension=yes

[MESSAGES CONTROL]
enable=all
disable=I,too-many-lines

[REPORTS]
#reports=yes
msg-template='{path}:{line}:{column}: {msg_id} ({symbol}) {msg}'

[BASIC]
good-names=i,j,c,_
include-naming-hint=yes

[FORMAT]
max-line-length=120

[VARIABLES]
allow-global-unused-variables=no
```

### Command used

```shell
pylint pylint_bug.py
```

### Pylint output

```python
************* Module pylint_bug
pylint_bug.py:1:0: W0612 (unused-variable) Unused variable '_HAS_DEFAULT_FACTORY'

------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 6.67/10, +0.00)
```

### Expected behavior

No errors.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

Windows 11 Home
10.0.22631 N/D Compilación 22631

**Comment by DanielNoord** [comment]
I can't reproduce.. Is `dataclasses` overwritten by a local module?

**Comment by DervishD** [comment]
Not that I know of, @DanielNoord but I'll check my system just in case, as soon as possible. During the weekend I think I can do a clean reinstall of Python on my system, removing all installed Python packages, just in case some of them is the culprit (they are all pretty standard, pylint, its dependencies and a few others) and I'll try to get my hands on another otherwise pristine system and I will test there, too.

Thanks for the interest!

**Comment by DervishD** [comment]
I could take a minute and I did a thorough cleanup of my system (to the best of my knowledge, of course, I can't be sure there are not any cached files left), performed a clean install of Python 3.12, a clean install of Pylint and I can reproduce the problem.

I don't know if I made clear that my config file (the one I posted) has `allow-global-unused-variables=no`, but this is intended, in order to produce W0612 errors for unused global variables, which is something I want for my code as a default. Still, this is a false positive.

So, since you cannot reproduce the problem, @DanielNoord but I can after a clean install, is there any way I can produce some debugging output for pylint so I can try to pinpoint the problem myself?

Thanks a lot in advance!

**Comment by Pierre-Sassoulas** [comment]
Hey @DervishD thank you opening an issue. Are you able to provide a dockerfile based on python:3.11 to reproduce ?

For example:
[3.11.6-bookworm](https://github.com/docker-library/python/blob/35d09c044857f7aef2bf24791027f3e3fe2c34dd/3.11/bookworm/Dockerfile)

Or if this is windows specific:
    [3.11.6-windowsservercore-ltsc2022](https://github.com/docker-library/python/blob/35d09c044857f7aef2bf24791027f3e3fe2c34dd/3.11/windows/windowsservercore-ltsc2022/Dockerfile)
    [3.11.6-windowsservercore-1809](https://github.com/docker-library/python/blob/35d09c044857f7aef2bf24791027f3e3fe2c34dd/3.11/windows/windowsservercore-1809/Dockerfile)

**Comment by DervishD** [comment]
@Pierre-Sassoulas I replied to your comment but somehow the reply got lost. I'll try to reproduce it: I have never used docker (or other containers), and I don't know how to create one, or a dockerfile for that matter. If you just need a dockerfile which reflects my system, I think I could provide one using the ones you point as base, but I don't know if it would be useful.

Last, but not least: can the dockerfile be based on python 3.12? I can reproduce the problem in that version and it is already installed in the two computers I use for development, the same ones I used to reproduce the problem.

While I try to learn how to write a dockerfile, is there any other way I can help with debugging?

Thanks A LOT for your reply and your interest!

**Comment by Pierre-Sassoulas** [comment]
Sorry, I should have said python 3.12, my mistake. The way to do that would be to install docker and then you can create an almost empty docker file

```dockerfile
FROM python:3.12
RUN pip3 install pylint==3.0.2
RUN ...
```
You can test what it does with ``docker run -it .`` (or docker exec to test directly inside it before creating the run). Then we will be able to reproduce using either the dockerfile you run with ``docker run`` or the list of command you used on the base docker file in ``docker exec``

**Comment by Pierre-Sassoulas** [comment]
Thank you for the tip and all your effort to reproduce @DervishD, I can reproduce with:
``a.py``
```python
import dataclasses

@dataclasses.dataclass
class SomeDataClass():
    ...
```
Command: ``pylint a.py --allow-global-unused-variables=no``

**Comment by DervishD** [comment]
Thanks to you, @Pierre-Sassoulas for your interest in this problem 😄 My pleasure, and if you finally need the dockerfile, just tell me and I'll try my best to produce one.

**Comment by DervishD** [comment]
I can confirm that it still happens in `Windows 11, python 3.12, pylint 3.2.4` .

**Comment by DervishD** [comment]
Tested again with `Windows 11`, `python 3.13.2` and `pylint 3.2.7`. I've tried to find the cause but since I'm not familiar with the code I haven't found anything, sorry.

This looks like a very obscure problem. It's not a HUGE problem, though, it is merely aesthetic, but it hurts me not being able to find the problem.

Thanks for the ongoing effort @Pierre-Sassoulas .

**Comment by LukeShu** [comment]
Confirming that I see this on GNU/Linux with:

```console
$ pylint --version
pylint 3.3.7
astroid 3.3.11
Python 3.13.5 (main, Jun 21 2025, 09:35:00) [GCC 15.1.1 20250425]
```

**Comment by LukeShu** [comment]
The problem is that `astroid/brain/brain_dataclasses.py` `dataclass_transform()` injects a `_HAS_DEFAULT_FACTORY` into `node.root().locals` without either
 - checking whether `_generate_dataclass_init()` will actually use it
 - doing something mark it as used

**Comment by LukeShu** [comment]
I suppose an alternative fix instead of checking whether it will be used or preemptively marking it would be to point the actual stdlib `dataclasses._HAS_DEFAULT_FACTORY`, instead of creating a mock of it in the current module. But that'd be relying on an implementation detail of the stdlib that maybe isn't safe for non-cpython implementations, or future cpython versions?

**Comment by Pierre-Sassoulas** [comment]
We can differentiate between cpython and pypy in the dataclass astroid brain. We'll see what we can do about terseness if a major python interpreter appear, when it come to it :)

**Comment by DervishD** [comment]
I've tested on my end and yes, it appears to be fixed. Thanks A LOT.
